Add support for custom hashers

mjclarke94 opened the pull request:
Attempts to close #32. Still needs documentation etc, but figured I'd get the initial PR up for code review.

I'm fairly new to rust so feel free to be brutal here! It's likely that I'm missing something obvious here.

I was looking into whether it would be plausible to also be generic over the map type (say, to allow `BTreeMap` instead of `HashMap` which might be handy when it comes to things like `k_most_common`, but was struggling to see how to make it be generic over both.

coriolinus commented:
Thanks for the contribution! It's likely to be a few days before I can properly review this--life is hectic right now--but I have seen it, and it's on my task queue.

mjclarke94 commented:
I'm installing off my fork for the time being so no rush on my end :) 
This is just the result of me getting carried away trying to optimise an advent of code solution!

coriolinus commented:
Good news! I'm finally back and able to review this, after "a few days" somehow turned into half a year.

Bad news: I made some changes, not to this pr. I think they'll make the PR cleaner, but it means there are now some merge conflicts.

Can't expect better behavior from anyone else than I can of myself. So! Let's get this fixed up and merged by 2026!

Commit:
feat: add support for custom hashers

Adds a third generic parameter S (defaulting to RandomState) to Counter,
allowing users to specify custom hash builders for the underlying HashMap.
This enables use cases like deterministic hashing or performance optimizations.

Key changes:
- Add generic parameter S = RandomState to Counter struct
- Update all trait implementations to include the S parameter
- Add BuildHasher bounds only where HashMap operations require them
- Add Default bounds only for constructors
- Include test demonstrating custom hasher usage

Closes #32

coriolinus reviewed (approved):
Thanks for the patience, and for this work!

govinda-kamath commented:
Looks like this broke the build with `serde` enabled. Fixed here: https://github.com/coriolinus/counter-rs/pull/54

QuadnucYard commented:
Any reason for the missing `<S>` in the impl block of `into_map` and `total`?
I just found that the custom hash surprisingly does not support these 😅

https://github.com/coriolinus/counter-rs/blob/e518825d6cda3b1648a6fa06f518b568f1b68020/src/lib.rs#L316-L345
